internal/pkg/receive/hashrings.go

```diff
@@ -145,7 +159,13 @@ func EndpointSliceListToEndpoints(converter EndpointConverter, eps discoveryv1.E
 		}
 	}
 	sort.Slice(endpoints, func(i, j int) bool {
-		return endpoints[i].Address < endpoints[j].Address
+		if endpoints[i].Address != "" && endpoints[j].Address != "" {
+			return endpoints[i].Address < endpoints[j].Address
+		}
+		if endpoints[i].CapnProtoAddress != "" && endpoints[j].CapnProtoAddress != "" {
```

Comment:
nit: Is the only difference between the `CapnProtoAddress` and the `Address` the port? If so perhaps we could simplify the logic for generating the endpoints.